apps/examples/testcase/le_tc/kernel/tc_umm_heap.c

```diff
@@ -134,25 +192,120 @@ static void tc_umm_heap_calloc(void)
 
-	for (alloc_tc_cnt = 0; alloc_tc_cnt < TEST_TIMES; alloc_tc_cnt++) {
-		for (alloc_cnt = 0; alloc_cnt < ALLOC_FREE_TIMES; alloc_cnt++) {
-			mem_ptr[alloc_cnt] = (int *)realloc(val, ALLOC_SIZE_VAL * sizeof(int));
-			TC_ASSERT_NEQ("realloc", mem_ptr[alloc_cnt], NULL);
+	for (n_test_iter = 0; n_test_iter < TEST_TIMES; n_test_iter++) {
+		/* Save mallinfo before test start */
+
+#ifdef CONFIG_CAN_PASS_STRUCTS
+		prev = mallinfo();
+#else
+		(void)mallinfo(&prev);
+#endif
+
+		/* Allocate memory */
+
+		for (n_alloc = 0; n_alloc < ALLOC_FREE_TIMES; n_alloc++) {
+			mem_ptr[n_alloc] = (int *)realloc(prev_mem, alloc_size);
+			TC_ASSERT_NEQ("realloc", mem_ptr[n_alloc], NULL);
 		}
-		heap = mm_get_heap(mem_ptr[alloc_cnt - 1]);
+
+		/* Verify allocation */
+
+#ifdef CONFIG_CAN_PASS_STRUCTS
+		cur = mallinfo();
+#else
+		(void)mallinfo(&cur);
+#endif
+		// Due to remain size, it could be greater than sizeof(int) + SIZEOF_MM_ALLOCNODE.
+		TC_ASSERT_EQ_CLEANUP("mallinfo", cur.uordblks - prev.uordblks, MEM_REQ_SIZE(alloc_size, ALLOC_FREE_TIMES),
+		                     mem_deallocate_func(mem_ptr, ALLOC_FREE_TIMES));
+		TC_ASSERT_EQ_CLEANUP("mallinfo", prev.fordblks - cur.fordblks, MEM_REQ_SIZE(alloc_size, ALLOC_FREE_TIMES),
+                             mem_deallocate_func(mem_ptr, ALLOC_FREE_TIMES));
+
+#ifdef CONFIG_DEBUG_MM_HEAPINFO
+		/* Verify allocation by heapinfo */
+
+		heap = mm_get_heap(mem_ptr[n_alloc - 1]);
 		TC_ASSERT_NEQ_CLEANUP("realloc", heap, NULL, mem_deallocate_func(mem_ptr, ALLOC_FREE_TIMES));
-		TC_ASSERT_EQ_ERROR_CLEANUP("realloc", heap->alloc_list[hash_pid].curr_alloc_size, TOTAL_ALLOC_SIZE, get_errno(), mem_deallocate_func(mem_ptr, ALLOC_FREE_TIMES));
+		TC_ASSERT_EQ_ERROR_CLEANUP("realloc", heap->alloc_list[hash_pid].curr_alloc_size,
+		          MEM_REQ_SIZE(alloc_size, ALLOC_FREE_TIMES), get_errno(), mem_deallocate_func(mem_ptr, ALLOC_FREE_TIMES));
+#endif
+
+		/* Free allocated memory */
+
 		mem_deallocate_func(mem_ptr, ALLOC_FREE_TIMES);
 
+		/* Verify freeing */
+
+#ifdef CONFIG_CAN_PASS_STRUCTS
+		cur = mallinfo();
+#else
+		(void)mallinfo(&cur);
+#endif
+		TC_ASSERT_EQ("mallinfo", cur.uordblks - prev.uordblks, 0);
+		TC_ASSERT_EQ("mallinfo", prev.fordblks - cur.fordblks, 0);
+
+#ifdef CONFIG_DEBUG_MM_HEAPINFO
+		/* Verify freeing by heapinfo */
+
 		TC_ASSERT_EQ("realloc", heap->alloc_list[hash_pid].curr_alloc_size, ALL_FREE);
+#endif
 	}
+
+	/* Relloc Free by size 0 */
+
+	mem_ptr[0] = (int *)malloc(10);
+	mem_ptr[1] = (int *)realloc(mem_ptr[0], 0);
+	TC_ASSERT_EQ_CLEANUP("realloc", mem_ptr[1], NULL, free(mem_ptr[0]));
+
+	/* Verify freeing */
+
+#ifdef CONFIG_CAN_PASS_STRUCTS
+	cur = mallinfo();
+#else
+	(void)mallinfo(&cur);
+#endif
+	TC_ASSERT_EQ("mallinfo", cur.uordblks - prev.uordblks, 0);
+	TC_ASSERT_EQ("mallinfo", prev.fordblks - cur.fordblks, 0);
+
+	/* Relloc Free by size 0 */
+	mem_ptr[0] = (int *)malloc(alloc_size);
+	alloc_size /= 2;
+	mem_ptr[1] = (int *)realloc(mem_ptr[0], alloc_size);
```

Comment:
I think it is better to check `mem_ptr[0]` is allocated well or not.
If `mem_ptr[0]` is NULL, then `mem_ptr[1]` can be also NULL, so this tc will be failed.